_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,244 @@
+
+Placing the Cruise Control metric reporter jar into the '/opt/kafka/libs/' directory of every Kafka broker allows Kafka to find the reporter at runtime. For the sake of simplicity, one can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:
+
+```Dockerfile
+FROM strimzi/kafka:0.13.0-kafka-2.3.0
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"
+
+USER root
+
+RUN yum -y install git java-1.8.0-openjdk-devel \
+    && yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+RUN cd cruise-control \
+    && ./gradlew jar :cruise-control-metric-reporter:jar \
+    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
+```
+Then building the image:
+```bash
+docker build . -t <kafka-image-name>
+```
+For the Cruise Control metric reporter to be activated at runtime, its class name must be listed in the 'metrics.reporters' field of the Kafka config. For a Strimzi Kafka cluster, one can do this by creating a Kafka custom resource which references the new Kafka 'image' built above and the 'com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter' class name in the 'metrics.reporters' field using the following 'kafka.yaml' file:
+```yaml
+apiVersion: kafka.strimzi.io/v0alpha1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    image: <kafka-image-name>
+    version: 2.3.0
+    replicas: 3
+    listeners:
+      plain: {}
+      tls: {}
+    config:
+      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
+      offsets.topic.replication.factor: 1
+      transaction.state.log.replication.factor: 1
+      transaction.state.log.min.isr: 1
+      log.message.format.version: "2.3"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+```
+Then applying the Kafka custom resource to the cluster:
+
+```bash
+kubectl apply -f kafka.yaml
+```
+The 'metrics.reporters' field can take a list of comma-separated metric reporter class names. At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.
```

Comment:
I missing here the part how CC gets these metrics. Can we add something like _CC will read these metrics from the designated Kafka topic and use them for balancing. How to configure the connection between CC and Kafka will be descibed later in one of the next parts of this blog post._

Reply:
At this point in the article the it's not yet been made clear the overall picture of what's going on. I think this whole metrics section could do with an outline at the start:

"""
Each broker is configured with a metric reporter. The metric reporter publishes a subset of the broker metrics to a special topic. Cruise Control consumes the metrics from this topic to learn about the load experienced by each broker in the cluster. 
"""

Then I think you can elaborate on each of those parts and the reader has a good enough mental model of the moving parts for it to all make sense.